src/arithmetic/algebraic_expression.c

```diff
@@ -305,95 +201,357 @@ void AlgebraicExpression_PrependTerm(AlgebraicExpression *ae, GrB_Matrix m, bool
+    }
+
+    // TODO: might be optimized with memcpy.
+    // Shift operands to the right, making room at the begining.
+    for(int i = ae->operand_count-1; i > 0 ; i--) {
+        ae->operands[i] = ae->operands[i-1];
+    }
+
+    ae->operands[0] = op;
+}
+
+static void _RemovePathFromGraph(QueryGraph *g, Edge **path) {
+    uint edge_count = array_len(path);
+    for(uint i = 0; i < edge_count; i++) {
+        Edge *e = path[i];
+        Node *src = e->src;
+        Node *dest = e->dest;
+
+        QueryGraph_RemoveEdge(g, e);
+        if(Node_EdgeCount(src) == 0) QueryGraph_RemoveNode(g, src);
+        if(Node_EdgeCount(dest) == 0) QueryGraph_RemoveNode(g, dest);
+    }
+}
+
+/* Determin the length of the longest path in the graph.
+ * Returns a list residing on the edge of the longest path. */
+static Node** _DeepestLevel(const QueryGraph *g, int *level) {
+    Node *s = g->nodes[0];
+    int l = BFS_LOWEST_LEVEL;
+    Node **leafs = BFS(g->nodes[0], &l);
+    Node *leaf = leafs[0];
+    array_free(leafs);
+
+    l = BFS_LOWEST_LEVEL;
+    leafs = BFS(leaf, &l);
+
+    *level = l;
+    return leafs;
+}
+
+/* Break down expression into sub expressions.
+ * considering referenced intermidate nodes and edges. */
+static AlgebraicExpression** _AlgebraicExpression_Intermidate_Expressions(
+    AlgebraicExpression *exp, Edge **path, const QueryGraph *g,
+    TrieMap *ref_entities) {
+
+    /* Allocating maximum number of expression possible. */
+    Edge *e = NULL;
+    int expIdx = 0;         // Sub expression index.
+    int operandIdx = 0;     // Index to currently inspected operand.
+    int pathLen = array_len(path);
+    AlgebraicExpression **expressions;
+
+    expressions = array_new(AlgebraicExpression*, exp->operand_count);
+    AlgebraicExpression *iexp = _AE_MUL(exp->operand_count);
+    iexp->operand_count = 0;
+    iexp->src_node = exp->src_node;
+    iexp->dest_node = exp->dest_node;
+    expressions = array_append(expressions, iexp);
+    expIdx++;
+
+    for(int i = 0; i < pathLen; i++) {
+        e = path[i];
+        /* If edge is referenced, set expression edge pointer. */
+        if(_referred_entity(e->alias, ref_entities)) iexp->edge = e;
+
+        /* If this is a variable length edge, which is not fixed in length
+         * remember edge length. */
+        if(Edge_VariableLength(e)) {
+            iexp->edge = e;
         }
 
-        if(exp->operand_count == 0) {
-            exp->src_node = src;
-            if(src->label) {
-                GrB_Matrix srcMat = Node_GetMatrix(src);
-                AlgebraicExpression_AppendTerm(exp, srcMat, false, false);
-            }
+        if(i == 0 && e->src->label) {
+            iexp->operands[iexp->operand_count++] = exp->operands[operandIdx++];
+        }
+
+        /* Expand fixed variable length edge */
+        unsigned int hops = (!Edge_VariableLength(e)) ? e->minHops : 1;
+        for(int i = 0; i < hops; i++) {
```

Comment:
Use `j`; this overwrites outer loop variable `i` at 292.

Reply:
👍